### PR TITLE
Adding fixes for signedness warnings in libsolidity

### PR DESCRIPTION
**Pull request by mijovic**

### CODE REVIEW

File: test/libsolidity/SMTCheckerJSONTest.cpp
Comment (hrkrshnn) [inline test/libsolidity/SMTCheckerJSONTest.cpp]: These changes can be avoided if we define
```
size_t start = location["start"].asUInt();
size_t end = location["end"].asUInt();
```

After that
```suggestion
			if (start >= preamble.size())
				start -= preamble.size();
			if (end >= preamble.size())
				end -= preamble.size();
```

File: libsolidity/codegen/ContractCompiler.cpp
Comment (hrkrshnn) [inline libsolidity/codegen/ContractCompiler.cpp]:
```suggestion
				solAssert(functionEntryLabel.data() <= std::numeric_limits<size_t>::max(), "");
```

**Review by hrkrshnn — COMMENTED** [review]
I don't see any problems. My suggestions are minor. However, a second review would be great!

File: test/libsolidity/SMTCheckerJSONTest.cpp
Comment (mijovic) [reply test/libsolidity/SMTCheckerJSONTest.cpp]: Thanks for suggestion, really made it much simpler

**Comment by leonardoalt** [comment]
Windows build is failing

**Comment by mijovic** [comment]
> Windows build is failing

Fixed

File: libsolidity/analysis/OverrideChecker.cpp
Comment (cameel) [inline libsolidity/analysis/OverrideChecker.cpp]: Maybe change the type of `m_graph.numNodes` to `size_t`? It would then require only one cast, in `visit()`.

File: libsolidity/codegen/CompilerContext.cpp
Comment (cameel) [inline libsolidity/codegen/CompilerContext.cpp]:
```suggestion
		setStackOffset(static_cast<int>(inArgs) + 1);
```

File: libsolidity/codegen/CompilerContext.cpp
Comment (cameel) [inline libsolidity/codegen/CompilerContext.cpp]:
```suggestion
	return static_cast<unsigned>(m_asm->deposit()) - _baseOffset - 1;
```

Comment (cameel) [inline libsolidity/codegen/CompilerContext.cpp]: Maybe also assert that `m_asm->deposit() > _baseOffset`?

File: libsolidity/codegen/CompilerContext.cpp
Comment (cameel) [inline libsolidity/codegen/CompilerContext.cpp]:
```suggestion
	return static_cast<unsigned>(m_asm->deposit()) - _offset - 1;
```

Comment (cameel) [inline libsolidity/codegen/CompilerContext.cpp]: Maybe also assert that `m_asm->deposit() > _baseOffset`?

File: libsolidity/codegen/CompilerContext.cpp
Comment (cameel) [inline libsolidity/codegen/CompilerContext.cpp]:
```suggestion
		auto stackDepth = static_cast<size_t>(distance(it, _localVariables.end()));
```

File: libsolidity/codegen/CompilerContext.cpp
Comment (cameel) [inline libsolidity/codegen/CompilerContext.cpp]: `stackDiff` is already `size_t`.
```suggestion
			_assembly.appendInstruction(dupInstruction(stackDiff));
```

File: libsolidity/codegen/CompilerUtils.cpp
Comment (cameel) [inline libsolidity/codegen/CompilerUtils.cpp]:
```suggestion
	m_context.adjustStackOffset(static_cast<int>(amount));
```

File: libsolidity/codegen/ContractCompiler.cpp
Comment (cameel) [inline libsolidity/codegen/ContractCompiler.cpp]:
```suggestion
			static_cast<int>(CompilerUtils::sizeOnStack(functionType->returnParameterTypes())) -
			static_cast<int>(CompilerUtils::sizeOnStack(functionType->parameterTypes())) -
```

File: libsolidity/codegen/ContractCompiler.cpp
Comment (cameel) [inline libsolidity/codegen/ContractCompiler.cpp]:
```suggestion
		m_context.adjustStackOffset(static_cast<int>(parametersSize) + 1);
```

File: libsolidity/codegen/ContractCompiler.cpp
Comment (cameel) [inline libsolidity/codegen/ContractCompiler.cpp]:
```suggestion
		stackLayout.push_back(static_cast<int>(i));
```

File: libsolidity/codegen/ContractCompiler.cpp
Comment (cameel) [inline libsolidity/codegen/ContractCompiler.cpp]:
```suggestion
	stackLayout.push_back(static_cast<int>(c_returnValuesSize)); // target of return address
```

File: libsolidity/codegen/ContractCompiler.cpp
Comment (cameel) [inline libsolidity/codegen/ContractCompiler.cpp]:
```suggestion
			m_context << swapInstruction(stackLayout.size() - static_cast<size_t>(stackLayout.back()) - 1);
```
We use `size_t` for this everywhere anyway.

File: libsolidity/codegen/ContractCompiler.cpp
Comment (cameel) [inline libsolidity/codegen/ContractCompiler.cpp]:
```suggestion
					unsigned stackDiff = static_cast<unsigned>(_assembly.stackHeight()) - m_context.baseStackOffsetOfVariable(*variable);
```

File: libsolidity/codegen/ContractCompiler.cpp
Comment (cameel) [inline libsolidity/codegen/ContractCompiler.cpp]:
```suggestion
			unsigned stackDiff = static_cast<unsigned>(_assembly.stackHeight()) - m_context.baseStackOffsetOfVariable(*variable) - 1;
```

File: libsolidity/codegen/ContractCompiler.cpp
Comment (cameel) [inline libsolidity/codegen/ContractCompiler.cpp]:
```suggestion
	m_context.setStackOffset(static_cast<int>(startStackHeight));
```

File: libsolidity/codegen/ExpressionCompiler.cpp
Comment (cameel) [inline libsolidity/codegen/ExpressionCompiler.cpp]:
```suggestion
	m_context.adjustStackOffset(1 + static_cast<int>(CompilerUtils::sizeOnStack(paramTypes)));
```

File: libsolidity/codegen/ExpressionCompiler.cpp
Comment (cameel) [inline libsolidity/codegen/ExpressionCompiler.cpp]:
```suggestion
			m_context.adjustStackOffset(static_cast<int>(returnParametersSize - parameterSize) - 1);
```

File: libsolidity/codegen/ExpressionCompiler.cpp
Comment (cameel) [inline libsolidity/codegen/ExpressionCompiler.cpp]:
```suggestion
				m_context.adjustStackOffset(static_cast<int>(arguments.at(1)->annotation().type->sizeOnStack()));
```

File: libsolidity/codegen/ExpressionCompiler.cpp
Comment (cameel) [inline libsolidity/codegen/ExpressionCompiler.cpp]:
```suggestion
	unsigned remainsSize =
```

File: libsolidity/formal/SMTEncoder.cpp
Comment (cameel) [inline libsolidity/formal/SMTEncoder.cpp]:
```suggestion
		auto trueIndex = static_cast<unsigned>(_indicesEndTrue.at(decl));
		auto falseIndex = static_cast<unsigned>(_indicesEndFalse.at(decl));
```

File: libsolidity/formal/SMTEncoder.cpp
Comment (cameel) [inline libsolidity/formal/SMTEncoder.cpp]:
```suggestion
		m_context.variable(*var.first)->setIndex(static_cast<unsigned>(var.second));
```

File: libsolidity/interface/CompilerStack.cpp
Comment (cameel) [inline libsolidity/interface/CompilerStack.cpp]:
```suggestion
	path.append(_path.begin() + static_cast<string::difference_type>(longestPrefix), _path.end());
```

File: test/libsolidity/SolidityExpressionCompiler.cpp
Comment (cameel) [inline test/libsolidity/SolidityExpressionCompiler.cpp]:
```suggestion
			context.adjustStackOffset(static_cast<int>(parametersSize));
```

**Review by cameel — COMMENTED** [review]
Only stylistic changes. Mostly explicit `static_cast` but also a few others.

File: libsolidity/analysis/OverrideChecker.cpp
Comment (mijovic) [reply libsolidity/analysis/OverrideChecker.cpp]: Makes sende, did like that

File: libsolidity/codegen/CompilerContext.cpp
Comment (mijovic) [reply libsolidity/codegen/CompilerContext.cpp]: Added

File: libsolidity/codegen/CompilerContext.cpp
Comment (mijovic) [reply libsolidity/codegen/CompilerContext.cpp]: Added

**Comment by mijovic** [comment]
@cameel I think I fixed all comments. Would be great to get another review round

**Comment by ekpyron** [comment]
There are still failing builds, though.

**Comment by ekpyron** [comment]
Might be something weird in some of the CI runs running on merge commits with develop, not on the PR branch, maybe you just have to rebase.

**Comment by mijovic** [comment]
Builds should be fixed now, let's see if CI will pass now

**Comment by leonardoalt** [comment]
Tests still failing, looks rather weird

**Comment by cameel** [comment]
@mijovic OK. I'm looking at it now.

File: libsolidity/codegen/CompilerContext.cpp
Comment (ekpyron) [inline libsolidity/codegen/CompilerContext.cpp:301]: Almost looks like this assertion is actually triggered in CI... can you add a mesage to it so that it becomes clearer?

File: libsolidity/codegen/CompilerContext.cpp
Comment (ekpyron) [inline libsolidity/codegen/CompilerContext.cpp]: Actually looks like this assertion may be triggered in CI, can you add a message to it?
(I first commented on the wrong one a few lines up - but a message for both doesn't hurt - we did even consider trying to actively get rid of the ``""`` assertions at some point :-) )

Comment (cameel) [reply libsolidity/codegen/CompilerContext.cpp]: `_baseOffset` >= `m_asm->deposit()` would be a bug, right?

The assertion wasn't there before but I suggested adding it since otherwise this might wrap around. But I don't know this code well. There's a small chance that this was the intention.

> we did even consider trying to actively get rid of the "" assertions at some point :-) )

Wouldn't really be that hard to add a variant without the message and gradually keep replacing them. Do we actually want this? I thought it was by design, to force you to add a message.

**Comment by cameel** [comment]
@mijovic I looked through all the new changes and everything seems fine to me. The only remaining issue is the failing assertion. But if it was already broken before this PR then I think leaving it as is here and fixing it in a separate PR would be fine too.

**Comment by ekpyron** [comment]
CI is *really* weird on this one... why doesn't it start building the linux builds? I was about to say that it's weird that we get that exception only on windows, but the circleci runs don't even start here apparently...

**Comment by mijovic** [comment]
> @mijovic I looked through all the new changes and everything seems fine to me. The only remaining issue is the failing assertion. But if it was already broken before this PR then I think leaving it as is here and fixing it in a separate PR would be fine too.

I will remove assertion and ope issue to track

**Review by cameel — APPROVED** [review]

**Pull request merged**
